numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1757,36 +1776,70 @@ unpack_bits(PyObject *input, int axis)
         unpack_init = 1;
     }
 
-    NPY_BEGIN_THREADS_THRESHOLDED(PyArray_DIM(new, axis));
+    count = PyArray_DIM(new, axis) * 8;
+    if (outdims[axis] > count) {
+        in_n = count / 8;
+        in_tail = 0;
+        out_pad = outdims[axis] - count;
+    }
+    else {
+        in_n = outdims[axis] / 8;
+        in_tail = outdims[axis] % 8;
+        out_pad = 0;
+        /* threshold will be zero if any dimension is zero, including count */
+        if (threshold) {
+            threshold *= 8 * in_n / count;
+        }
+    }
 
-    n_in = PyArray_DIM(new, axis);
     in_stride = PyArray_STRIDE(new, axis);
     out_stride = PyArray_STRIDE(out, axis);
 
+    NPY_BEGIN_THREADS_THRESHOLDED(threshold);
```

Comment:
Sorry, that this is causing so much unnecessary trouble... I am still a bit confused. Firstly, can't count be below zero here? Secondly, what is the advantage compared to using `PyArray_Size(out)` maybe `PyArray_Size(out)` that seemed to me like a reasonable and painless estimation of the total effort?

Reply:
You're absolutely right. My thinking was that the number of elements we want to unpack is the significant number here, but `PyArray_Size(out) / 8` is a much better way than the abomination I've created. Will fix ASAP.

Not that it matters any more, but `count` can't be negative here. That's tested for explicitly.

Reply:
You're absolutely right. My thinking was that the number of elements we want to unpack is the significant number here, but `PyArray_Size(out) / 8` is a much better way than the abomination I've created. Will fix ASAP.

Not that it matters any more, but `count` can't be negative here. That's tested for explicitly.

Reply:
You're absolutely right. My thinking was that the number of elements we want to unpack is the significant number here, but `PyArray_Size(out) / 8` is a much better way than the abomination I've created. Will fix ASAP.

Not that it matters any more, but `count` can't be negative here. That's tested for explicitly.